pipelines/training_audit.py

```
@@ -255,9 +263,251 @@ def hidden_thought_paths(value, path=""):
+        for descriptor in opened:
+            os.close(descriptor)
+
+
+def _enclosing_marker_root(run_dir: Path, path: Path) -> Path | None:
+    """Return the nearest marker-mode factory enclosing ``path``, if any."""
+
+    current = path.parent
+    while True:
+        if marker_mode_path(current) is not None:
+            return current
+        if current == run_dir or current.parent == current:
```

Ruff (0.16.3)

[warning] Consider merging multiple comparisons: `current in (run_dir, current.parent)`. Use a `set` if the elements are hashable.

Merge multiple comparisons

(PLR1714)